apollo-federation/src/source_aware/federated_query_graph/path_tree.rs

```diff
@@ -17,18 +22,270 @@ pub(crate) struct FederatedPathTree {
+    // rather than breadth-first.
+    fn from_paths_inner<'a, EdgeIter, PathIter>(
+        graph: Arc<FederatedQueryGraph>,
+        node: NodeIndex,
+        paths: PathIter,
+        remapped_condition_ids: &mut HashMap<ConditionResolutionId, ConditionResolutionId>,
+    ) -> Result<Self, FederationError>
+    where
+        EdgeIter: Iterator<Item = &'a graph_path::Edge>,
+        PathIter: Iterator<Item = (EdgeIter, Option<Arc<SelectionSet>>)>,
+    {
+        let mut condition_resolutions_at_node = IndexMap::new();
+        let mut source_entering_condition_resolutions_at_node = IndexMap::new();
+
+        struct ByUniqueEdge<'a, EdgeIter>
+        where
+            EdgeIter: Iterator<Item = &'a graph_path::Edge>,
+        {
+            target_node: NodeIndex,
+            by_unique_trigger:
+                IndexMap<Option<Arc<OperationPathElement>>, PathTreeChildInputs<'a, EdgeIter>>,
+        }
+
+        struct PathTreeChildInputs<'a, EdgeIter>
+        where
+            EdgeIter: Iterator<Item = &'a graph_path::Edge>,
+        {
+            self_condition_resolutions_for_edge:
+                IndexMap<SelfConditionIndex, ConditionResolutionId>,
+            sub_paths_and_selections: Vec<(EdgeIter, Option<Arc<SelectionSet>>)>,
+        }
+
+        let mut merged = IndexMap::new();
+        for (mut graph_path_iter, selection) in paths {
+            let Some(edge) = graph_path_iter.next() else {
+                continue;
+            };
+
+            merge_condition_resolutions(
+                &mut source_entering_condition_resolutions_at_node,
+                &edge.source_entering_condition_resolutions_at_head,
+                remapped_condition_ids,
+            );
+            merge_condition_resolutions(
+                &mut condition_resolutions_at_node,
+                &edge.condition_resolutions_at_head,
+                remapped_condition_ids,
+            );
+
+            // Marginally inefficient: we look up edge endpoints even if the edge already exists.
+            // This is to make the ? error propagate.
+            // XXX(@goto-bus-stop): Should we store the endpoint on the graph path edges? Then this
+            // code would not need to return a result. The query graph is not mutable at this point so
+            // it can not get out of sync.
+            let for_edge = merged.entry(edge.edge).or_insert(ByUniqueEdge {
+                target_node: if let Some(edge) = &edge.edge {
+                    let (_source, target) = graph.edge_endpoints(*edge)?;
+                    target
+                } else {
+                    // For a "None" edge, stay on the same node
+                    node
+                },
+                by_unique_trigger: IndexMap::new(),
+            });
+
+            match for_edge
+                .by_unique_trigger
+                .entry(edge.operation_element.clone())
+            {
```

Comment:
This handles paths having potentially multiple different triggers for the same edge. What does that mean in practice? Two different fields can never share an edge in the query graph, right?